### PR TITLE
Compressed cursors fix problem with little buffers

### PR DESCRIPTION
**Pull request by eolivelli**
Fix a problem that happens when the compressed data is smaller than the original data.

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -3429,35 +3430,44 @@ private ByteBuf compressDataIfNeeded(ByteBuf data, LedgerHandle lh) {
...
         if (pulsarCursorInfoCompression != null) {
             String pulsarCursorInfoCompressionString = new String(pulsarCursorInfoCompression);
+            log.info("Ledger {} compression {} decompressing {} bytes, full {}",
+                    lh.getId(), pulsarCursorInfoCompressionString, data.length, ByteBufUtil.prettyHexDump(Unpooled.wrappedBuffer(data)));
             ByteArrayInputStream input = new ByteArrayInputStream(data);
             DataInputStream dataInputStream = new DataInputStream(input);
             try {
                 int uncompressedSize = dataInputStream.readInt();
-                byte[] compressedData = dataInputStream.readNBytes(uncompressedSize);
+                byte[] compressedData = dataInputStream.readAllBytes();
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3451]: This is actually the fix !

**Review by nicoloboschi — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -3429,35 +3430,44 @@ private ByteBuf compressDataIfNeeded(ByteBuf data, LedgerHandle lh) {
             int ratio = (int) (compressedSize * 100.0 / uncompressedSize);
             log.info("[{}] Cursor {} Compressed data size {} bytes (with {}, original size {} bytes, ratio {}%)",
                     ledger.getName(), name, compressedSize, pulsarCursorInfoCompressionString, uncompressedSize, ratio);
+            log.info("Compressed data {}, full {}", ByteBufUtil.prettyHexDump(encode), ByteBufUtil.prettyHexDump(result));
```
Comment (nicoloboschi) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3433]: debug?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 3]
```diff
@@ -3410,6 +3410,7 @@ private ByteBuf compressDataIfNeeded(ByteBuf data, LedgerHandle lh) {
         }
 
         try {
+            log.info("Original data {}", ByteBufUtil.prettyHexDump(data));
```
Comment (nicoloboschi) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3413]: debug ?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 4]
```diff
@@ -3429,35 +3430,44 @@ private ByteBuf compressDataIfNeeded(ByteBuf data, LedgerHandle lh) {
...
         }
     }
 
-    private static byte[] decompressDataIfNeeded(byte[] data, LedgerHandle lh) {
+    private byte[] decompressDataIfNeeded(byte[] data, LedgerHandle lh) {
         byte[] pulsarCursorInfoCompression =
                 lh.getCustomMetadata().get(METADATA_PROPERTY_CURSOR_COMPRESSION_TYPE);
         if (pulsarCursorInfoCompression != null) {
             String pulsarCursorInfoCompressionString = new String(pulsarCursorInfoCompression);
+            log.info("Ledger {} compression {} decompressing {} bytes, full {}",
```
Comment (nicoloboschi) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3445]: debug?

**Pull request merged**